### PR TITLE
Adds support for direct and recurring donations

### PR DESCRIPTION
**Pull request by NejcZdovc**
Resolves https://github.com/brave/brave-browser/issues/925
Resolves https://github.com/brave/brave-browser/issues/1550

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
(for developers only)
Uncomment this line https://github.com/brave-intl/bat-native-ledger/blob/master/src/bat_client.cc#L183 to trigger contribution after 5min

1.
- clean profile
- enable rewards
- claim grant
- visit some sites (wait 10s on them)
- wait for contribution to happen

2.
- clean profile
- enable rewards
- claim grant
- visit some sites and add them to recurring donation via panel and site banner
- wait for contribution to happen

3.
- clean profile
- enable rewards
- claim grant
- visit some sites (wait 10s on them)
- visit some sites and add them to recurring donation via panel and site banner
- wait for both contribution to happen

4.
- clean profile
- enable rewards
- claim grant
- visit site
- click on panel, donate now and pick amount
- make sure that direct donation happens

## Notes

1. Currently you can't do recurring donation and ac at the same time. You will get error like this `Error: the signed and unsigned transactions differed`. This will be fixed on Monday (15/10) on the server

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: browser/ui/webui/brave_donate_ui.cc [hunk 1]
```diff
@@ -164,6 +174,43 @@ void RewardsDonateDOMHandler::OnWalletProperties(
...
+
+void RewardsDonateDOMHandler::OnRecurringDonationUpdated(brave_rewards::RewardsService* rewards_service,
+                                                   const brave_rewards::ContentSiteList list) {
+   if (web_ui()->CanCallJavascript()) {
+    auto publishers = std::make_unique<base::ListValue>();
+    for (auto const& item : list) {
+      publishers->AppendString(item.id);
+    }
+
+    web_ui()->CallJavascriptFunctionUnsafe("brave_rewards_donate.recurringDonations", *publishers);
```
Comment (diracdeltas) [inline browser/ui/webui/brave_donate_ui.cc]: can we follow this suggestion instead: https://cs.chromium.org/chromium/src/content/public/browser/web_ui.h?q=CallJavascriptFunctionUnsafe&sq=package:chromium&g=0&l=90-93

**Comment by bbondy** [comment]
master: 19e7f54847ef4d7e5720174dadc155f3ceab4a4b
0.56.x: 9ed26c208dc0231c46aab50045c5ae58b9ef029c
0.55.x: 52ced8fc5d2115c5b57f081046451edf29674fd5

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 2]
```diff
@@ -690,6 +695,26 @@ void RewardsServiceImpl::LoadPublisherInfoList(
     uint32_t limit,
     ledger::PublisherInfoFilter filter,
     ledger::GetPublisherInfoListCallback callback) {
+  auto now = base::Time::Now();
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: why do we have two separate methods for this? What if the code calling this method already set month and year to something else? They are going to get unexpected results. If you are going to predefine filter values then you shouldn't have the filter as a parameter

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 3]
```diff
@@ -1310,4 +1335,195 @@ brave_rewards::PublisherBanner RewardsServiceImpl::GetPublisherBanner(const std:
...
+  PublisherInfoDatabase* backend) {
+  if (backend && backend->InsertContributionInfo(info))
+    return true;
+
+  return false;
+}
+
+void RewardsServiceImpl::OnContributionInfoSaved(const ledger::PUBLISHER_CATEGORY category, bool success) {
+  if (success && category == ledger::PUBLISHER_CATEGORY::DIRECT_DONATION) {
+    TipsUpdated();
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: I don't think we should be doing this in RewardsServiceImpl. If the UI needs tip info when the contribution info is saved then we should have an observer method for it and let the webui call TipsUpdated. We're basically putting UI logic in the RewardsServiceImpl by doing things like this and I don't think it's a good idea to do that. If you want a single observer per profile to call this then you can create a new one, but it should be part of the UI code, not part of the service code

Comment (bridiver) [reply components/brave_rewards/browser/rewards_service_impl.cc]: the RewardsServiceImpl should just do the basics and let observers handle anything else that needs to happen when events occur

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 2 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/rewards_service_impl.cc]: fixed in https://github.com/brave/brave-core/pull/1029

**Pull request merged**